src/main/java/seedu/address/model/ModelManager.java

```diff
@@ -79,6 +88,54 @@ public synchronized void addTask(Task person) {
         indicateTaskBookChanged();
     }
 
+    //=============for undo and redo===================================
+    @Override
+    public Command undo() throws EmptyStackException {
+        Commit lastCommit = commitStack.pop();
+        redoables.push(new Commit(lastCommit.getCommand(), new TaskBook(getTaskBook())));
+        resetData(lastCommit.getTaskBook());
+        Command undoneAction = lastCommit.getCommand();
+        return undoneAction;
+    }
+
+    @Override
+    public void discardRecentCommit() {
+        commitStack.pop();
+    }
+
+    @Override
+    public void resetRedoables() {
+        redoables = new Stack<Commit>();
+    }
+
+    @Override
+    public void recordState(Command command) {
+        commitStack.push(new Commit(command, new TaskBook(getTaskBook())));
+    }
+
+    public Commit snapshot(Command command) {
```

Comment:
there must be some commit error on my part, i was supposed  to unstage that hunk